### DISCUSSION TITLE
How to create three or more users tables in adonis5?

### DISCUSSION THREAD

**Discussion by tiagocdr** [adonisjs/core#2971] [Help]
I'm new to adonis 5 and I'm building an app with different user types (customers, employees, admins, etc...), the authentication module just lets you create and use for authentication one user, I've been reading on documentation and reading a lot of discussions about the subject but I cant find anything conclusive. I have tried different approaches but I'm just hitting dead ends in all of them, here I'll list a few of them: 

1. Permissions
 - The issue I have with this approach is that eventually I find myself adding too many clauses and and chained if statements making everything far too complicated as I think it could be, as well as giving me a problem of organization in regards of how the user data is all mixed up in the database. 
 
 2. Inheritance
 - This seems like the solution I'm most hopeful about but I can't seem to figure out how to implement it, Reading on the documentation I stumbled about the helper adonis provides to simplify mixins in typescript but I'm not sure how organize this inheritance, if you could provide some examples about creating models with mixins and registering the users I would really appreciate it.
 
 3. One user with many profiles 
  - This is the method I'm currently using but so far, even tho its working in a certain way, seems very fragile specially if the process fails during the registration, I have one user and at the moment of signing up I create the user and then separately create the profile within the same request, but if the profile creation fails the user still remains in the database and I have to go and check for it and delete the user, it gets complicated and makes everything go out of "sync",  still if this is the best approach to take I'm happy to read some better ideas on how to implement this. 
  
4. Custom Provider.
- Every-time you start looking for ways of implementing or solving this issue you will find this as pretty much the solution, many of the cases and examples I've seen are either from Adonis4, which I haven't used at all and seems very different from the new typescript version, in which from what I've seen documentation and discussions about, It doesn't seem to be what I'm looking for/need or they are just either unanswered or half answered questions without much further explanation. 

Open to any solutions for best/easiest/most secure way of implementing this multiple users models, hopefully this can be the post devs will finally say 'this is it!' for it, as I've seen many posts on the web asking the same question being unanswered. 

In advance thank you for taking the time to read this.

**Comment by arthur-er** [comment]
1. You could use permissions and write a role based ACL with only one user model and only one profile model and use bouncer / middleware to authorize users based on role

2. You could also use a mixin to lookup permissions in the database (like spatie/laravel-permission for laravel), and authorize them against a bouncer policy
 
3. You could wrap the user and profile creation in a transaction and rollback if one of them fails (https://docs.adonisjs.com/guides/database/transactions#using-transactions-with-lucid-models)

4. I don't think that writing a provider is what you are looking for. Currently the docs are lacking but you can look at some provider implementations from the core team or from the comunity and see how it works

**Reply by arthur-er** [reply]
Also if that is the case you can register more auth guards for other user models, just look at the auth config and auth contract to see how it works
